webnn/clamp.https.any.js

```diff
@@ -0,0 +1,158 @@
+// META: title=test WebNN API clamp operation
+// META: global=window,dedicatedworker
+// META: script=./resources/utils.js
+// META: timeout=long
+
+'use strict';
+
+// https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-clamp
+
+const inputData = [
+  0.58585083,  1.1363881,   0.67161655,  -0.9741674,  -1.6196846,
+  0.572627,    1.9026182,   -0.7756641,  -0.18808974, -1.0357478,
+  1.1778295,   -2.305167,   -2.2636602,  0.3750199,   -0.08234365,
+  -0.47962302, -0.3010948,  0.5369879,   -0.413804,   -1.096925,
+  -0.9273629,  0.88833886,  -0.52474195, -1.3852776,  0.10217833,
+  0.50499475,  1.3289608,   0.21790339,  -0.65971124, 0.47400787,
+  0.7271749,   -0.03890531, -0.04459939, 0.2601329,   -0.06985649,
+  0.2501139,   -1.0219133,  -1.1504377,  -0.83611137, 0.64221096,
+  0.25879756,  1.040239,    -0.18669093, -1.1436414,  1.1445535,
+  -0.01876706, 1.283455,    0.59794647,  2.1886187,   -0.21977298,
+  0.90072393,  0.8913641,   -0.55512637, -0.17248231, -1.4617383,
+  -1.5487962,  0.1265688,   0.7930071,   0.63802403,  0.3400246,
+];
+// expected data by clamping input data within a range [-1, 1]
+const expected = [
```

Comment:
Hi @fdwr , based on last discussion in WebML CG meeting, here input value of `Number` type (**float64**) will be casted to float32 before computing, see below line#75

```js
const inputs = {'x': new Float32Array(inputValue)};
```
while each element of this `expected` Array is also of `Number` type (**float64**), we should use expected data of **float32** for float32 test case, so it needs cast output value from **float64** to **float32**, on this purpose, I need update line#87

```js
assert_array_approx_equals_ulp(outputs.y, expected, ULPTolerance.float32.clamp, 'float32');
```
to 
```js
assert_array_approx_equals_ulp(outputs.y, new Float32Array(expected), ULPTolerance.float32.clamp, 'float32');
```

is it right?  Thanks.

Reply:
Yes, it's important for the *inputs* to be the same data type, so both reference baseline engine and execution provider implementations start out from the same input value. Otherwise additional error would immediately be introduced that isn't fair to the GPU implementation because precision was already lost before the GPU even saw the input.

For *output*, we can compare the GPU's float32 output to the reference engine's downcasted (float64 -> float32) output, but it should also be okay to compare the upcasted GPU output (float32 -> float64) with the reference engine, so long as you shift the ULP values accordingly (float64 has 52 fraction bits, while float32 has 23 fraction bits, yielding a shift of <<29). One advantage to the latter approach is that you could see more precise final output values, without losing any additional precision from rounding down the baseline output float64->float32. However, the former is simpler and close enough, and it's what I've always done.

Do you know if the float64 to `Float32Array` cast rounds to nearest even or toward zero? I couldn't tell in my quick search.

Reply:
Sorry, the above solution of expected data for test case seems not correct, from beginning we need pass the **float32** input to webnn-baseline then get/generate output data of **float32**, then use those input/output (float32) for float32 WPT test cases,  would be this right? Thanks.

Reply:
🤔 How are you generating the arrays of expected output data?

Reply:
> How are you generating the arrays of expected output data?

We've already implemented some operations functions of [webnn-baseline](https://github.com/webmachinelearning/webnn-baseline) repo by pure JavaScript without depending on 3rd party libs. We could (1) generate and save input data by invoking `Math.random()` in JavaScript, (2) run operation of [webnn-baseline](https://github.com/webmachinelearning/webnn-baseline) with feeding these input data, then (3) get and save output data as expected data. These input value and output value are of float64, for float32 purpose, we need cast input data to float32 in step (2), right?

Reply:
@huningxin @wchao1115 Would you please take a look about float32 baseline? Thanks.

Reply:
@fdwr 

> Do you know if the float64 to Float32Array cast rounds to nearest even or toward zero? I couldn't tell in my quick search.

It should cast rounds to nearest even.  As the `NumericToRawBytes` for Float32 type in [ECMA-262](https://tc39.es/ecma262/multipage/structured-data.html#sec-numerictorawbytes), it says "the result of converting value to IEEE 754-2019 binary32 format using roundTiesToEven mode".

Reply:
@BruceDai 

>  Would you please take a look about float32 baseline? Thanks.

As @fdwr suggested, we probably need to convert the inputs to float32 before feed to `webnn-baselin`, e.g., by using `Float32Array`. The `webnn-baseline` would still use double-precision calculation and generate double-precision outputs. Then outputs should be convert to float32 again as expected results.